avoid unnecessary COUNT queries when exporting

PetrDlouhy opened the pull request:
**Problem**

Fix for #1714 

**Solution**

Described at #1714 

**Acceptance Criteria**

I didn't yet write tests and documentation. I would like to get some feedback first.

matthewhegarty commented:
It sounds interesting, thanks for picking it up.
Does the FakePaginator break pagination when we have larger datasets?

PetrDlouhy commented:
@matthewhegarty I didn't know about that functionality, but according to my tests it seems the chunking works OK. If you look at the code (https://github.com/django-import-export/django-import-export/blob/b63ddd6148c786573838fbfc0e30131dfdef3a71/import_export/resources.py#L1087) it seems that `django-import-export` creates it's own Paginator which is independent on the Paginator created by `ModelAdmin`.
Also I hope the chunking is already covered by unit tests.

Here is comparison of the called queries after this change:
![Snímek obrazovky_2023-12-14_12-39-00](https://github.com/django-import-export/django-import-export/assets/156755/905d30e0-fdcb-469f-8ea0-8bab31e4f50f)
And before this change:
![Snímek obrazovky_2023-12-14_12-38-13](https://github.com/django-import-export/django-import-export/assets/156755/9d6846fc-ae12-494e-a096-cbacb9a8a639)

In both cases there is query for each chunk, but in the first case the COUNT queries are missing.


BTW: I was confused for a while how the chunking works until I figured out from the code that it is used only if the query contains `prefetch_related`, which is something I don't see mentioned in the docs: https://django-import-export.readthedocs.io/en/stable/installation.html#import-export-chunk-size

PetrDlouhy commented:
@matthewhegarty I refactored the code, added some comments, changelog mention and test for number of queries.
Is there anything more that might be needed?

matthewhegarty commented on import_export/admin.py:
```diff
@@ -741,7 +760,12 @@ def get_export_queryset(self, request):
-        cl = ChangeList(**changelist_kwargs)
+
+        # Temporarily set to False to avoid unnecessary COUNT queries.
+        with temp_attr(self, "show_full_result_count", False):
```
I think this is fine, but I'm not sure if it will have any side effects.  How about we wrap this block in a setting called `IMPORT_EXPORT_DISABLE_EXPORT_COUNT`, then we can document the reasons for this, and users can opt into disabling the check if they like.

matthewhegarty commented on tests/core/tests/test_admin_integration.py:
```diff
@@ -680,7 +680,10 @@ def test_export(self):
         }
         date_str = datetime.now().strftime("%Y-%m-%d")
-        response = self.client.post("/admin/core/book/export/", data)
+        with self.assertNumQueries(
```
I would prefer a more robust check to prove that `show_full_result_count` and `get_paginator` were not called - perhaps they could be mocked?

I think that counting the number of queries is prone to failing in future when Django is updated.

matthewhegarty reviewed (changes requested):
Looks good - thanks for submitting - I have a couple of comments.

matthewhegarty commented:
> BTW: I was confused for a while how the chunking works until I figured out from the code that it is used only if the query contains prefetch_related, which is something I don't see mentioned in the docs

If you wanted to update the docs for this to make it clearer, it would be much appreciated.

PetrDlouhy commented:
@matthewhegarty I think this change would be beneficial to all users of `django-import-export` and should be transparent.
I would rather address all concerns about side effects than hiding this behind some configuration.

I have looked at it once more and I found better approach to this. It is no longer modifying the `ModelAdmin`, but I rather overriding the `ChangeList.get_results()`. I hope that is even less intrusive than the original approach.
Now the only scenario I can think of this would affect users is if they were relying on the `ChangeList.get_results()` being called. I think that there is only very slight chance that somebody uses such thing. I think it would be enough to mention this in the changelog.


In my production environment I use `django-cachalot`, so I didn't notice this before, but Django already calls the `get_queryset()` method during initialization of the `ChangeList`. This means that `get_queryset()` is called twice and the SQL queries too.
I have fixed this and even further reduced number of queries called.

I will try to improve the testing and also look at #974 if there is some connection to this.

PetrDlouhy commented:
@matthewhegarty Now I have improved the testing, now it ensures number of calls to the paginator and get_queryset.
I have left the `assertNumQueries` checks, because I think it would be good to ensure the number of queries will not bloat spontaneously. I can remove that, if you think contrary.

I have also written some more detailed information about the change to the changelog.

matthewhegarty reviewed (changes requested):
Thanks for updating.  I prefer this implementation to the previous version.  Happy to have this enabled for all users.

Just some minor changes requested re wording.

matthewhegarty commented on docs/changelog.rst:
```diff
@@ -5,7 +5,8 @@ Changelog
 - Nothing changed yet.
-
+- Remove unnecessary ChangeList queries to speed up export via Admin UI (#1715)
+  This is done by overriding `ChangeList` from `ModelAdmin.get_changelist().get_results()` with stub method.
```
Please can you remove this sentence (link to PR is enough)

matthewhegarty reviewed (changes requested):
Thanks - one last change requested then I will merge 👍

matthewhegarty commented:
Thanks - much appreciated.
